dangerzone/isolation_provider/base.py

```diff
@@ -142,10 +174,16 @@ def doc_to_pixels(
             n_pages = read_int(p.stdout)
             if n_pages == 0 or n_pages > errors.MAX_PAGES:
                 raise errors.MaxPagesException()
-            percentage_per_page = 49.0 / n_pages
+            step = 100 / n_pages
+
+            safe_doc = fitz.Document()
 
             for page in range(1, n_pages + 1):
-                text = f"Converting page {page}/{n_pages} to pixels"
+                searchable = "searchable " if ocr_lang else ""
+                text = (
+                    f"Converting page {page}/{n_pages} from pixels to {searchable}PDF"
+                )
+                percentage += step
                 self.print_progress(document, False, text, percentage)
```

Comment:
I think so, yeah. Any issue you see with this approach?

Reply:
Sorry, my previous message wasn't clear :-) I'm thinking about passing a different context to `print_progress` so that it actually prints the progress itself. Here the message is computed outside of this function, and I'm thinking we could do this inside the method instead.

Reply:
Hm, something like passing a context that holds the total number of pages and the current page, so that `print_progress()` can construct the proper "Converting page..." message? I guess we can, but we'll still need to call `print_progress() with an arbitrary message, e.g., in case of errors. Is there another benefit that we gain from this?

Reply:
First: this would work without changes, and we can consider this discussion as cosmetic. If you prefer to ship things as they are right now, it would be fine by me.

With that being said, my thinking here follows what you've stated: rather than computing the message and the percentage outside of the `self.print_progress()` method, we could actually do it there. The parameters to the method would change to something like this:

```python
def print_progress(
        self, document: Document, page: int, n_pages: int, finished: Optional[bool] = None
    ) -> None:
```

Following this lead, it's probably best to add a `print_error` method for errors rather than having `print_progress` do both, and we could also factorize some of the color handling in a third method.

To me it's a matter of responsibility as to who is deciding on the message to be printed, I believe this results in the code being easier to read.

Reply:
Nope, that's ok. I was mostly trying to understand where you're going with this, to be honest. The code and the explanation made things much more concrete for me, thanks! I think I have something interesting I can come up with, let me draft something.

Reply:
Ok, I'm back with a proposal: https://github.com/freedomofpress/dangerzone/commit/fff7be753547d1c651a62f6b90fa3dae91095a88 (branch: `625-host-stream-4`)

The idea is this: the isolation provider is a shared object for all conversions. We don't have a conversion context (as you have alluded to in your comment) that we can refer to for a document. If we had something like that, what would it look like?

One thing you pointed out in this thread is that it could be in charge of printing messages. But what about the percentage of the conversion? Shouldn't it track it as well? And if it could do that, shouldn't it be able to track a few more things:
1. What is the state of the conversion for a specific document?
2. What are the conversion parameters?
3. What is the estimated remaining time for the conversion?

So, this is what I did with this commit. I factored out the things that we implicitly do in the isolation provider, in a provider agnostic `ConversionCtx` class. What I'd like to know is if what I created here is in the right direction or not, based on your comments.

If yes, I can proceed with some tests, and including it in this PR. If not, then let's reconvene and I'll amend the design.

Reply:
Thanks for proposing this. After some more thoughts, I believe this is a welcome addition, and going to the right direction. Feel free to merge with it, or do it in another PR so we can merge this one quicker!

Reply:
Cool! I'll probably send this as a different PR then, so that I can merge this first. This will unblock two of your PRs who are waiting, and it will give me time to work on it while you're out.